ie_serving/server/predict_utils.py

```diff
@@ -61,32 +62,46 @@ def prepare_input_data(models, model_name, version, data, rest):
                 message = str(e)
                 logger.debug("PREDICT prepare_input_data make_ndarray error: "
                              "{}".format(message))
-                return True, message, None, code
+                return True, message, code
         else:
             tensor_input = np.asarray(data[requested_input_blob])
+        # Validate shape if shape not in auto mode
+        if models[model_name].engines[version].shape_info.mode != ShapeMode.AUTO:
+            shape_required_in_model = models[model_name].engines[version] \
+                .input_tensors[tensor_name].shape
+
+            # For reshapable models check all dimensions,
+            # for non-reshapable, check all starting from the second (ommit
+            # batch size)
+            if models[model_name].engines[version].shape_info.mode == \
+                    ShapeMode.DISABLED:
+                starting_dim = 1
+            else:
+                starting_dim = 0
+
+            # check requested shape and model shape
+            if shape_required_in_model[starting_dim:] != list(
+                    tensor_input.shape)[starting_dim:]:
+                code = statusCodes['invalid_arg'][request_type]
+                message = INVALID_SHAPE.format(list(tensor_input.shape),
+                                               shape_required_in_model)
+                logger.debug("PREDICT error: {}".format(message))
+                return True, message, code
+
+            # check if input batch size match the model only if not auto mode
+            if models[model_name].engines[version].batching_info.mode != \
+                BatchingMode.AUTO and shape_required_in_model[0] != \
+                    tensor_input.shape[0]:
+                code = statusCodes['invalid_arg'][request_type]
+                message = INVALID_BATCHSIZE.format(
+                    tensor_input.shape[0],
+                    models[model_name].engines[
+                        version].batching_info.batch_size)
+                logger.debug("PREDICT error,Invalid batchsize:{}".format(message))
+                return True, message, code

```

Comment:
We can delete first return value `True/False` and instead use `code` variable, if is it `None` keep going if not, return message and code